src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -164,43 +165,89 @@ public boolean partitionExists(final String topic, final String partition) throw
                 .anyMatch(partition::equals);
     }
 
-    @Override
-    public void syncPostBatch(final String topicId, final List<BatchItem> batch) throws EventPublishingException {
-        final CountDownLatch done = new CountDownLatch(batch.size());
-
-        for (final BatchItem item : batch) {
-            Preconditions.checkNotNull(item.getPartition(),
-                    "BatchItem partition can't be null at the moment of publishing!");
-            final ProducerRecord<String, String> record = new ProducerRecord<>(topicId,
-                    toKafkaPartition(item.getPartition()), item.getPartition(), item.getEvent().toString());
-
-            item.setStep(EventPublishingStep.PUBLISHING);
+    private static CompletableFuture<Exception> publishItem(
+            final Producer<String, String> producer, final String topicId, final BatchItem item)
+            throws EventPublishingException {
+        try {
+            final CompletableFuture<Exception> result = new CompletableFuture<>();
+            final ProducerRecord<String, String> kafkaRecord = new ProducerRecord<>(
+                    topicId,
+                    toKafkaPartition(item.getPartition()),
+                    item.getPartition(),
+                    item.getEvent().toString());
+
+            producer.send(kafkaRecord, ((metadata, exception) -> {
+                if (null != exception) {
+                    item.setStep(EventPublishingStep.FAILED_PUBLISH);
+                    item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+                    result.complete(exception);
+                } else {
+                    item.setStep(EventPublishingStep.PUBLISHED);
+                    item.updateStatusAndDetail(EventPublishingStatus.SUBMITTED, "");
+                    result.complete(null);
+                }
+            }));
+            return result;
+        } catch (final InterruptException e) {
```

Comment:
ok, it is not there, it is Exception from kafka, which they rethrow without interrupting current thread back.